File: web/unsplash.py
# Standard library imports
import json

# Third party imports
import requests

# Local imports
from web import config, geolocate


class UnsplashException(Exception):
	pass


def _send_request(endpoint: str, params: dict) -> dict:
	headers = {
		'Accept-Version': 'v1',
		'Authorization': 'Client-ID {}'.format(config.UNSPLASH_ACCESS_KEY)
	}
	resp = requests.get(
		'https://api.unsplash.com{}'.format(endpoint),
		params=params,
		headers=headers
	)
	resp.raise_for_status()
	return json.loads(resp.text)


def random(tags: list) -> dict:
	params = {
		'query': 'mountain',
		'orientation': 'landscape',
		'count': 30
	}
	resp = _send_request(
		'/photos/random',
		params
	)
	for r in resp:
		latitude = r['location']['position']['latitude']
		longitude = r['location']['position']['longitude']
		if latitude and longitude:
			location = geolocate.locate(latitude, longitude)
			if location:
				return {
					'id': r['id'],
					'urls': {
						'full': r['urls']['full'],
						'regular': r['urls']['regular']
					},
					'author': {
						'name': '{} {}'.format(r['user']['first_name'], r['user']['last_name']),
						'instagram': r['user']['instagram_username']
					},
					'location': location
				}
				return r

	raise UnsplashException('No images found with coordinates.')

File: src/lib/server/unsplash.ts
import { type AssetFull, createApi } from 'unsplash-js';
import { env } from '$env/dynamic/private';

if (!env.UNSPLASH_ACCESS_KEY) throw new Error('UNSPLASH_ACCESS_KEY is not set');

const unsplash = createApi({ accessKey: env.UNSPLASH_ACCESS_KEY });

interface UnsplashImage {
  id: string;
  urls: {
    full: string;
    regular: string;
  };
  author: {
    name: string;
    instagram: string | null;
  };
  location: {
    name: string | null;
    latitude: number | null;
    longitude: number | null;
  };
}

const getRandom = async (tags: string[]): Promise<UnsplashImage> => {
  const { data, error } = await unsplash.GET('/photos/random', {
    params: {
      query: {
        query: tags[Math.floor(Math.random() * tags.length)],
        orientation: 'landscape',
        count: 30,
      },
    },
  });

  if (error || data == null) {
    throw new Error('Unexpected response from Unsplash');
  }

  // Random endpoint returns a single photo or an array depending on count parameter
  // Handle both cases just in case
  const photos = (Array.isArray(data) ? data : [data]) as AssetFull[];
  const raw = photos.find((r) => r.location?.position?.latitude != null);
  if (!raw) throw new Error('No images found with coordinates');

  return {
    id: raw.id,
    urls: {
      full: raw.urls.full,
      regular: raw.urls.regular,
    },
    author: {
      name: raw.user.name,
      instagram: raw.user.social.instagram_username,
    },
    location: {
      name: raw.location.name,
      latitude: raw.location.position.latitude,
      longitude: raw.location.position.longitude,
    },
  };
};

export const UnsplashAPI = {
  getRandom,
};
